src/NumericalAlgorithms/RootFinding/GslMultiRoot.hpp

```diff
@@ -0,0 +1,364 @@
+  return GSL_SUCCESS;
+}
+}  // end namespace detail
+
+/*!
+ *  \ingroup NumericalAlgorithmsGroup
+ *  \brief The exception thrown when gsl_multiroot fails to converge.
+ */
+class rootfinding_exception : public std::exception {
+  const char* what() const noexcept override { return "gsl_multiroot failed."; }
+};
+
+/*!
+ *  \ingroup NumericalAlgorithmsGroup
+ *  \brief The different options for the verbosity of gsl_multiroot.
+ */
+enum class Verbosity {
+  /// Do not print anything
+  Silent,
+  /// Print only "success" or "failed" on termination
+  Quiet,
+  /// Print final functions values on termination
+  Verbose,
+  /// Print function values on every iteration
+  Debug
+};
+
+/*!
+ *  \ingroup NumericalAlgorithmsGroup
+ *  \brief The different options for the rootfinding method of gsl_multiroot.
+ *
+ *  These methods require the function to have a jacobian provided: Hybrids,
+ *  Hybrid, and Newton. These methods do not require a provided jacobian:
+ *  Hybrids, Hybrid, DNewton.
+ */
+enum class Method {
+  /// Hybrid of Newton's method along with
+  /// following the gradient direction.
+  /// Note: Sometimes Hybrids works only
+  /// with the TestResidual stopping
+  /// condition.
+  Hybrids,
+  /// "Unscaled version of Hybrids that
+  /// uses a spherical trust region," see
+  /// GSL documentation for more details.
+  Hybrid,
+  /// Modification on Newton's method
+  /// that improves global convergence
+  Newton
+};
+
+/*!
+ *  \ingroup NumericalAlgorithmsGroup
+ *  \brief The different options for the convergence criterion of gsl_multiroot.
+ */
+enum class StoppingCondition {
+  /// Requires an absolute tolerance as
+  /// well as a relative tolerance.
+  /// See GSL documentation for
+  /// gsl_multiroot_test_delta.
+  TestDelta,
+  /// Requires only an absolute
+  /// tolerance. See GSL documentation
+  /// for gsl_multiroot_test_residual.
+  TestResidual
+};
+
+namespace detail {
+CREATE_IS_CALLABLE(jacobian)
+
+template <typename SolverType, typename SolverAlloc, typename SolverSet,
+          typename SolverIterate, typename SolverFree, size_t Dim,
+          typename Function>
+std::array<double, Dim> gsl_multiroot_impl(
+    Function& f, const std::array<double, Dim>& initial_guess,
+    double absolute_tolerance, size_t maximum_iterations,
+    double relative_tolerance, Verbosity verbosity,
+    double maximum_absolute_tolerance, SolverType solver_type,
+    StoppingCondition condition, SolverAlloc solver_alloc, SolverSet solver_set,
+    SolverIterate solver_iterate, SolverFree solver_free) {
+  // Supply gsl_root with the initial guess:
+  gsl_vector* const gsl_root = gsl_vector_alloc(Dim);
+  gsl_vector_set_with_std_array(gsl_root, initial_guess);
+  auto* solver = solver_alloc(solver_type, Dim);
+  solver_set(solver, &f, gsl_root);
+
+  // Take iterations:
+  if (verbosity == Verbosity::Verbose or verbosity == Verbosity::Debug) {
+    Parallel::printf("\nAttempting a root find...\n");
```

Comment:
If you really want to print this something that actually mentions the algorithm and parameters would be very useful.